code/game/machinery/turrets/turrets.dm

```diff
@@ -0,0 +1,225 @@
+// Base type for rewritten turrets, designed to hopefully be cleaner and more fun to use and fight against.
+// Use the subtypes for 'real' turrets.
+
+/obj/machinery/turret
+	name = "don't use me"
+	icon = 'icons/obj/turrets.dmi'
+	icon_state = "turret_barrel"
+	anchored = TRUE
+	idle_power_usage = 500
+	active_power_usage = 60000 // This determines how fast the internal cell gets charged.
+
+	transform_animate_time = 0.2 SECONDS
+
+	// Visuals.
+	var/image/turret_stand = null
+	var/image/turret_ray = null
+	var/ray_color = "#FFFFFFFF" // Color of the ray, changed by the FSM when switching states.
+
+	// Sounds
+	var/turn_on_sound = null // Played when turret goes from off to on.
+	var/turn_off_sound = null // The above, in reverse.
+
+	// Shooting
+	var/obj/item/gun/installed_gun = null // Instance of the gun inside the turret.
```

Comment:
A pattern I've found useful has been something like `installed_gun = /obj/item/gun` and then in init `if(ispath(installed_gun)) installed_gun = new installed_gun(src)`. Saves an extra largely unused var declaration.